python/benchmark/benchmark/bench_dbscan.py

```diff
@@ -0,0 +1,304 @@
+            Note: this column is assumed to be of pyspark sql 'array' type.
+        prediction_col
+            Name of prediction column
+
+        Returns
+        -------
+        float
+            The computed silhoutte score.
+
+        """
+        from sklearn.metrics import silhouette_score
+
+        sc = transformed_df.rdd.context
+
+        pdf: pd.DataFrame = transformed_df.toPandas()
+        features_pdf = pdf.drop(columns=[prediction_col])
+        prediction_pdf = pdf[prediction_col]
+
+        features_np = np.stack(features_pdf.to_numpy().squeeze())
+        prediction_np = prediction_pdf.to_numpy()
+
+        return silhouette_score(features_np, prediction_np)
+
+    def run_once(
+        self,
+        spark: SparkSession,
+        train_df: DataFrame,
+        features_col: Union[str, List[str]],
+        transform_df: Optional[DataFrame],
+        label_name: Optional[str],
+    ) -> Dict[str, Any]:
+        num_gpus = self.args.num_gpus
+        num_cpus = self.args.num_cpus
+        no_cache = self.args.no_cache
+        train_path = self.args.train_path
+
+        func_start_time = time.time()
+
+        first_col = train_df.dtypes[0][0]
+        first_col_type = train_df.dtypes[0][1]
+        is_array_col = True if "array" in first_col_type else False
+        is_vector_col = True if "vector" in first_col_type else False
+        is_single_col = is_array_col or is_vector_col
+        if not is_single_col:
+            input_cols = [c for c in train_df.schema.names]
+        output_col = "cluster_idx"
+
+        if num_gpus > 0:
+            from spark_rapids_ml.clustering import DBSCAN
+
+            assert num_cpus <= 0
+            if not no_cache:
+
+                def gpu_cache_df(df: DataFrame) -> DataFrame:
+                    df = df.repartition(num_gpus).cache()
+                    df.count()
+                    return df
+
+                train_df, prepare_time = with_benchmark(
+                    "prepare dataset", lambda: gpu_cache_df(train_df)
+                )
+
+            params = self.class_params
+            print(f"Passing {params} to DBSCAN")
+
+            gpu_estimator = DBSCAN(
+                num_workers=num_gpus, verbose=self.args.verbose, **params
+            )
+
+            if is_single_col:
+                gpu_estimator = gpu_estimator.setFeaturesCol(first_col)
+            else:
+                gpu_estimator = gpu_estimator.setFeaturesCols(input_cols)
+
+            gpu_model, fit_time = with_benchmark(
+                "gpu fit", lambda: gpu_estimator.fit(train_df)
+            )
+
+            # count doesn't trigger compute so do something not too compute intensive
+            transformed_df, transform_time = with_benchmark(
+                "gpu transform",
+                lambda: gpu_model.setPredictionCol(output_col).transform(train_df),
+            )
+
+            _, extra_transform_time = with_benchmark(
+                "gpu transform extra",
```

Comment:
Better to give this and above break downs in transform time more informative names corresponding to the major compute steps expected to actually occur in the respective cases (given sparks lazy compute model).